src/client/api/restful/send_tagged_data.c

```diff
@@ -0,0 +1,182 @@
+  cJSON* payload = NULL;
+
+  // create message object
+  if ((msg_obj = cJSON_CreateObject()) == NULL) {
+    printf("[%s:%d] creating message object failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // Add NULL network id
+  if (!cJSON_AddNullToObject(msg_obj, JSON_KEY_NET_ID)) {
+    printf("[%s:%d] creating network ID failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // add parents
+  if (utarray_to_json_string_array(tips->u.tips, msg_obj, JSON_KEY_PARENT_IDS) != JSON_OK) {
+    printf("[%s:%d] adding tips array to message object failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // add nonce
+  if (!cJSON_AddNullToObject(msg_obj, JSON_KEY_NONCE)) {
+    printf("[%s:%d] creating nonce failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // create payload object
+  if ((payload = cJSON_CreateObject()) == NULL) {
+    printf("[%s:%d] creating payload object failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // add type to payload
+  if (!cJSON_AddNumberToObject(payload, JSON_KEY_TYPE, CORE_MESSAGE_PAYLOAD_TAGGED)) {
+    printf("[%s:%d] adding type to payload failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // add tag
+  // max tag length is 64
+  char tag_str[BIN_TO_HEX_STR_BYTES(64)] = {0};
+  if (bin_2_hex(tag, tag_len, tag_str, sizeof(tag_str)) != 0) {
+    printf("[%s:%d] bin to hex tag conversion failed\n", __func__, __LINE__);
+    goto end;
+  }
+  if (!cJSON_AddStringToObject(payload, JSON_KEY_TAG, tag_str)) {
+    printf("[%s:%d] adding tag to payload failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // data
+  if (data) {
+    char* data_str = malloc(BIN_TO_HEX_STR_BYTES(data_len));
+    if (!data_str) {
+      printf("[%s:%d] OOM\n", __func__, __LINE__);
+      goto end;
+    }
+    if (bin_2_hex(data, data_len, data_str, BIN_TO_HEX_STR_BYTES(data_len)) != 0) {
+      printf("[%s:%d] bin to hex data conversion failed\n", __func__, __LINE__);
+      free(data_str);
+      goto end;
+    }
+    if (!cJSON_AddStringToObject(payload, JSON_KEY_DATA, data_str)) {
+      printf("[%s:%d] adding tag data failed\n", __func__, __LINE__);
+      free(data_str);
+      goto end;
+    }
+    free(data_str);
+  } else {
+    // add a null data to tagged data
+    if (!cJSON_AddNullToObject(payload, JSON_KEY_DATA)) {
+      printf("[%s:%d] adding null data payload failed\n", __func__, __LINE__);
+      goto end;
+    }
+  }
+
+  // add payload to message
+  if (!cJSON_AddItemToObject(msg_obj, JSON_KEY_PAYLOAD, payload)) {
+    printf("[%s:%d] adding payload failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // json object to json string
+  char* msg_str = NULL;
+  if ((msg_str = cJSON_PrintUnformatted(msg_obj)) == NULL) {
+    printf("[%s:%d] converting json to string failed\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // put json string into byte_buf_t
+  byte_buf_t* json_data = byte_buf_new();
+  json_data->data = (byte_t*)msg_str;
+  json_data->cap = json_data->len = strlen(msg_str) + 1;
```

Comment:
You may use a function `byte_buf_new_with_data` to do the same.

Reply:
Done.